### ISSUE TITLE
No attribute completions for instance of a decorated class

### ISSUE THREAD

**Issue by eugenhu** [davidhalter/jedi#1607]
When I run the following,
```python
import jedi

code = """from typing import TypeVar, Type

T = TypeVar('T', Type)

def my_decorator(x: T) -> T:
    return x

@my_decorator
class MyClass:
    my_attribute = 123

def my_function(x: MyClass) -> int:
    return x."""

script = jedi.Script(code)
completion = script.complete(13, 13)

print(completion)
```
(trying to find completions after `return x.`), I get
```
[]
```
whereas I would expect jedi to suggest `my_attribute`, and it seems strange that there are no completions at all.

If I change `T = TypeVar('T', Type)` to just `T = TypeVar('T')`, then the output looks about right
```
[<Completion: my_attribute>, <Completion: __annotations__>, <Completion: __class__>, <Completion: __delattr__>, <Completion: __dict__>, <Completion: __dir__>, <Completion: __doc__>, <Completion: __eq__>, <Completion: __format__>, <Completion: __getattribute__>, <Completion: __hash__>, <Completion: __init__>, <Completion: __init_subclass__>, <Completion: __module__>, <Completion: __ne__>, <Completion: __new__>, <Completion: __reduce__>, <Completion: __reduce_ex__>, <Completion: __repr__>, <Completion: __setattr__>, <Completion: __sizeof__>, <Completion: __slots__>, <Completion: __str__>]
```
However `T = TypeVar('T', bound=Type)` also does not work.

I am using Python 3.8.2 and I've tried this both on Jedi v0.17.0 and the current latest master.

**Comment by davidhalter** [comment]
Decorators are not real

The problem is that what you're doing is basically:

```
def my_decorator(x: T<type>) -> T<type>:
    return x
```

There is absolutely no guarantee that the return type is `MyClass`, it could as well be `str`.

To avoid this you will probably need to do this:

```
def my_decorator(x: Type[T]) -> Type[T]:
    return x
```

In general typing of decorators is pretty annoying, especially for functions.

**Comment by eugenhu** [comment]
I see, yeah I think I get it, using
```
def my_decorator(x: Type[T]) -> Type[T]:
    return x
```
instead works.

Thanks.

**Comment by eugenhu** [comment]
Hi, sorry to bring this issue back up again, but I've spent some time thinking about this and I think I've come up with a 'better' example than the initial one I gave. (I got a bit mixed up with what `TypeVar('T', Type)` meant and I agree that this just means the inferred return type of `my_decorator()` will always just be `Type`, I should've had `TypeVar('T', bound=Type)` instead, like the example below.)

```python
import jedi

code = """from typing import TypeVar, Type

T = TypeVar('T', bound=Type)

class MyClass:
    my_attribute = 123

def my_decorator(x: T) -> T:
    return x

x = MyClass
y = my_decorator(x)

y.my_"""

script = jedi.Script(code)
completion = script.complete(14, 5)

print(completion)
```
Output:
```
[]
```

Using `reveal_type()` in pyright shows
```
Type of "x" is "Type[MyClass]"
```
and
```
Type of "y" is "Type[MyClass]"
```

Using `script.infer(11, 0)` returns
```
[<Name full_name='__main__.MyClass', description='class MyClass'>]
```
and `script.infer(12, 0)` returns
```
[<Name full_name='typing.Type', description='instance Type'>]
```

This problem doesn't really have anything to do with decorators so the issue title is not very accurate in hindsight, but that was what I had come across in my own code.

If `T=TypeVar('T', bound=Type)` is changed to just `T=TypeVar('T')`, then Jedi suggests the `my_attribute` completion.

**Comment by PeterJCLaw** [comment]
I suspect that this issue is fundamentally the same as #1425 -- that Jedi currently doesn't fully preserve the _values_ which are passed through an annotated decorator (only their types, which in this case is `Type` rather than `Type[MyClass]`).

It does feel like Jedi should be able to handle this case though, as I think it does have a way to represent `Type[MyClass]`.
#1425 is a bit more complicated due to the lack of a representation for _function_ types within the type domain (which likely needs to be distinct from representing _`Callable`_).

@eugenhu I realise this isn't an ideal solution, however as a workaround have you tried removing the annotations from the decorator?

**Comment by davidhalter** [comment]
@PeterJCLaw Why do you think that Jedi should handle this? The `bound=` change doesn't really change anything IMO.

Note that if you manage to change this somehow and decorators work "better" in Jedi, I'm welcome to that change, even though I personally think that it's not "correct" :).

**Comment by PeterJCLaw** [comment]
I agree that the `bound=` probably isn't something which there's a lot of value in Jedi observing -- rather Jedi should probably treat `T = TypeVar('T', bound=<something>)` the same as `T = TypeVar('T')`.

After all the `bound` aspect is just applying an upper bound on the types from which the types passed to an argument which is typed as `T` can "inherit" from. ("inherit" in quotes because it's not quite inheritance when things like `Callable` and `Protocol` are in play). It's probably easier for Jedi to let a type checker handle the checking and just accept whatever is passed in.

I agree that the original code here (which used `T = TypeVar('T', Type)`) can legitimately have the decorator treat the passed-in value as being `Type` and let a type checker correct any errant usage.
In that case the `Type` is not an upper bound but instead is a strict limit on exactly the types which the `T` can take. (Aside: passing a single value like that is also invalid, as it means the type isn't really generic in any sense -- it only has a single possible type).

**Comment by PeterJCLaw** [comment]
Removing the decorator aspect of this, I think the crux of the issue is that Jedi is treating `TypeVar('T', bound=Type)` as being like `TypeVar('T', Type)` and not like `TypeVar('T')`, which I believe is incorrect (and certainly disagrees with `mypy` and other type checkers). Unfortunately `TypeVar` has some rather different meanings for what are rather similar spellings.

Essentially, in the following code (as @eugenhu's last post mentions), `Other` should be treated as being the same as `MyClass`. This happens without the `bound=Type`, but not when it's present.

``` python
from typing import TypeVar, Type

T = TypeVar('T', bound=Type)
# T = TypeVar('T')

class MyClass:
    def my_method(self):
        pass

def passthrough(x: T) -> T:
    return x

MyClass.

Other = passthrough(MyClass)

Other.
```

(I'm testing this using `./sith.py run infer foo.py 17 3` and commenting out different lines)

**Comment by PeterJCLaw** [comment]
I've been having a look into this. The challenge to making this work is that when inside the function, the type of `x` is also `T` and in that scenario we _do_ want to treat it like `Type`.

**Comment by eugenhu** [comment]
> @eugenhu I realise this isn't an ideal solution, however as a workaround have you tried removing the annotations from the decorator?

Removing the annotations also works but I've been using
```python
def my_decorator(x: Type[T]) -> Type[T]:
    return x
```
at the moment.

I've tried a few other test cases as well
```python
import jedi

code = """from typing import TypeVar, Type, Any, Union

class A:
    pass

class AA(A):
    pass

class AAA(AA):
    pass

class B:
    pass

T = TypeVar('T', bound=AA)

def passthrough(x: T) -> T:
    return x

x = AAA()
y = passthrough(x)"""

script = jedi.Script(code)

print(script.infer(20, 0))
print(script.infer(21, 0))
```
Here the output is
```
[<Name full_name='__main__.AAA', description='instance AAA'>]
[<Name full_name='__main__.AAA', description='instance AAA'>]
```

Replacing with `x = A()` or `x = B()`, the inferred type of `y` has description 'instance A' and 'instance B' so it seems like the bound on `T` is ignored. But if we have `x = A` or `x = B`, then the inferred type of `y` is 'instance AA' in both cases.

Could be something to do with bounded type variables and passing classes as function arguments.

Replacing with `T = TypeVar('T', bound=Any)` seems to behave the same as `T = TypeVar('T')`, but `T = TypeVar('T', bound=Union[Any, AA])` acts like `T = TypeVar('T', bound=AA)` for the different values of `x` I tried (bound is ignored for instances, but `y` has type 'instance AA' when `x` is a class).

**Comment by PeterJCLaw** [comment]
Yup, I found something similar when creating test cases.

This is complicated by Jedi having multiple different routes towards understanding the code. The logic is approximately:
- if there are annotations and the annotations return a non-empty result, use that
- otherwise inspect the function implementation

The annotations handling doesn't itself directly check the bound, so if there's a useful value for `T` from that analysis then that's what's used.
The main gotcha here is that when passing a type (i.e: `MyClass` rather than an instance `MyClass()`) then the annotations analysis looks for the type of that and gets `type` before generating an arbitrary instance of that type for the rest of Jedi to use. Unfortunately the conversion of `MyClass` -> `type` loses the class (I suspect it should become `Type[MyClass]`) so the conversion back ends up with an empty value.

Since that's empty the implementation gets inspected, at which point the bound is checked (which is correct for the value of `x` within the function scope) and that is assumed to be the value that gets returned.

I started playing with a new wrapper class which Jedi's [`TypeVar`](https://github.com/davidhalter/jedi/blob/master/jedi/inference/gradual/type_var.py#L51) could use in its `infer_type_vars` that would encapsulate this. That had promising results, but I'm not sure it's quite the right approach -- as noted above I think instead we need to have a representation of `Type[MyClass]` there.

I'm not likely to have time this weekend to look into this more, but I can put up the test cases I'd put together if that would be useful.

**Comment by davidhalter** [comment]
+1 for `Type[MyClass]`. IMO that's the right concept (however I don't think `typing.Type` is the same concept as `Type` that I'm using here)

> as being like TypeVar('T', Type) and not like TypeVar('T'), which I believe is incorrect (and certainly disagrees with mypy and other type checkers)

I don't think it's incorrect (as in the typing PEP 484 says something different). I think that's exactly why `typing.Type[T]` was created in the first place. The correct type of a class is `type` and not something like `type<T>` (at least in pure Python if you call `int.__class__`).

The reason why mypy and other type checkers are working around this is probably for practical reasons. It's simply more practical to have `T` be what you guys propose: `Type[MyClass]`. When instantiated, this will be `MyClass` again. The same is obviously also true for functions and IMO I would fix this in the same go.

If we are talking about fixes for this, we could probably just wrap all Ts that are functions/classes something like this:

```
class TypeVar(BaseTypingValue):   # Modification in inference/gradual/type_var
    def infer_type_vars(self, value_set):
        def iterate():
            for v in value_set:
                cls = v.py__class__()
                if cls.is_function() or cls.is_class():
                    cls = TypeWrapper(cls, v)
                yield cls

        annotation_name = self.py__name__()
        return {annotation_name: iterate()}

class TypeWrapper(ValueWrapper):
    def __init__(self, wrapped_value, original_value):
        super().__init__(wrapped_value)
        self._original_value = self._original_value

    def execute_annotation(self):
        return ValueSet({self._original_value})
```

Obviously no idea if this works well, it would obviously need some testing, but I guess this would work pretty well (and IMO makes it pretty clear that functions and classes are exceptions in a way that get treated differently in annotations).

**Comment by PeterJCLaw** [comment]
Yup, that's roughly what I had in mind here. I've adapted your code and added some test cases: #1613.

**Comment by davidhalter** [comment]
Fixed by #1613
